Log receipt of deprecated or unsupported requests

gavin-norman-sociomantic opened the pull request.

Commit:
Remove unnecessary `if`s

Previously, it was possible to create a RequestInfo instance with an empty
name field. This is not now possible, so any conditions checking for an
empty name can be removed.

gavin-norman-sociomantic commented:
WIP implementation. Needs testing in a node. Blocked on having a neo node compatible with swarm v5.x.x.

Commit:
Add support for specifying request timing stats in the request struct

The `addHandler` method of `RequestMap` currently accepts a bool which
determines whether timing stats are gathered for the request being added.
All other request properties are specified via fields in the request
struct, rather than by arguments passed to `addHandler`. It is now
possible to also specify the timing stats option in the same way, by
including a field `const bool timing = ?;` in the request struct. If this
field is present, the value of the argument passed to `addHandler` is
ignored.

Commit:
Add support for marking a request as scheduled for removal (deprecated)

Commit:
Log and track stats on requests that are scheduled for removal

gavin-norman-sociomantic commented:
Tested (in dmqnode), updated, and unblocked.

Commit:
Add stats log metric tracking usage of requests scheduled for removal

gavin-norman-sociomantic commented:
Updated with fixed stats log tests.

nemanja-boric-sociomantic commented on src/swarm/node/request/RequestStats.d:
```diff
@@ -745,4 +745,320 @@ public class RequestStats
+                if ( this.request == rhs.request && this.client == rhs.client )
+                    return 0;
+                else
+                    return 1;
```
`return this.request != rhs.request || this.client != rhs.client` if you wish.

nemanja-boric-sociomantic commented on src/swarm/node/request/RequestStats.d:
```diff
@@ -745,4 +745,320 @@ public class RequestStats
+            // Warn the first time a client sends an old request version.
+            if ( !info.active_this_hour(now) )
+                logger.warn("Client '{}' on {}:{} sent an old version of request '{}'",
+                    client, addr.address_bytes, addr.port, rq);
```
I think the comment here should be made that this is triggered only once per hour per client per request, not only the first time (as far as I can see).

gavin-norman-sociomantic replied:
Yes, that's correct :+1:

gavin-norman-sociomantic commented:
Updated:
* Addressed Nemanja's points.
* Refactored the flag into `recent_activity`, which has a more intuitive function (as opposed to being a flag that just says "method X did Y").

nemanja-boric-sociomantic commented on src/swarm/neo/node/ConnectionHandler.d:
```diff
@@ -109,6 +109,12 @@ class ConnectionHandler : IConnectionHandler
+            /// will handle it, but will log a warning.
+            // (Deliberately not named as a deprecation to avoid messing up
+            // searches for that keyword.)
+            bool scheduled_for_removal;
```
Release notes missing.

nemanja-boric-sociomantic commented on src/swarm/node/request/RequestStats.d:
```diff
@@ -745,4 +745,327 @@ public class RequestStats
+        {
+            /// Indicates whether any recent activity has occurred for this
+            /// client/request pair. (We use this flag to essentially remove
+            /// items from the map (`client_request_info, below`), rather than
```
The closing backtick is at the wrong position.

nemanja-boric-sociomantic commented on src/swarm/node/request/RequestStats.d:
```diff
@@ -745,4 +745,327 @@ public class RequestStats
+            if ( info is null )
+            {
+                this.client_request_info[ClientRequest(client.dup, rq.dup)]
+                    = ClientRequestInfo.init;
```
I was worried that this opens an attack window, but the allocation happens only once per client per request :+1:.

gavin-norman-sociomantic replied:
Right.

nemanja-boric-sociomantic commented on relnotes/schedule-for-removal.feature.md:
```diff
@@ -0,0 +1,17 @@
+
+In this way, it is possible to tell when an older request version is no longer
+in use, rather than having to guess.
+
```
Ah, here's the missing release note, but it should also say how the request is marked to be scheduled for removal.

nemanja-boric-sociomantic commented:
Small notes about release notes and a small typo, but LGTM overall.

gavin-norman-sociomantic commented:
Updated and rebased.
